File: app/src/main/java/com/sorcererxw/matthiasheiderichphotography/util/ScheduleManager.java
package com.sorcererxw.matthiasheiderichphotography.util;

import android.content.Context;
import android.widget.Toast;

import com.sorcererxw.matthiasheiderichphotography.MHApp;
import com.sorcererxw.matthiasheiderichphotography.services.WallpaperChangeService;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import timber.log.Timber;

/**
 * @description:
 * @author: Sorcerer
 * @date: 2016/12/17
 */

public class ScheduleManager {
    private ScheduledExecutorService mScheduledExecutorService =
            Executors.newSingleThreadScheduledExecutor();

    private Context mContext;

    public ScheduleManager(Context context) {
        mContext = context;
    }

    public void shutdown() {
        mScheduledExecutorService.shutdown();
        mScheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
    }

    public void startAutoChangeWallpaperIfNeeded() {
        if (MHApp.getInstance().getPrefs().getAutoRotateEnable().getValue()) {
            long delay = MHApp.getInstance().getPrefs().getAutoRotateTime().getValue();
            scheduleChangeWallpaperService(delay);
        }
    }

    public void scheduleChangeWallpaperService(long delay) {
        shutdown();
        schedule(mChangeWallpaperRunnable, delay, TimeUnit.MILLISECONDS);
    }

    public void schedule(Runnable runnable, long delay, TimeUnit timeUnit) {
        shutdown();
        mScheduledExecutorService
                .scheduleAtFixedRate(runnable, delay, delay, timeUnit);
    }

    private Runnable mChangeWallpaperRunnable = new Runnable() {
        @Override
        public void run() {
            Timber.d("run");
            WallpaperChangeService.changeWallpaper(mContext);
        }
    };
}

File: app/src/main/java/com/sorcererxw/matthiasheiderichphotography/util/ScheduleManager.kt
package com.sorcererxw.matthiasheiderichphotography.util

import android.content.Context
import com.sorcererxw.matthiasheiderichphotography.services.WallpaperChangeService
import timber.log.Timber
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

/**
 * @description:
 * @author: Sorcerer
 * @date: 2016/12/17
 */

class ScheduleManager(private val mContext: Context) {
    private var mScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()

    private val mChangeWallpaperRunnable = Runnable {
        Timber.d("run")
        WallpaperChangeService.changeWallpaper(mContext)
    }

    fun shutdown() {
        mScheduledExecutorService.shutdown()
        mScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    }

    fun startAutoChangeWallpaperIfNeeded() {
        if (AppPref.getInstance(mContext).autoRotateEnable.get()) {
            val delay = AppPref.getInstance(mContext).autoRotateTime.get()
            scheduleChangeWallpaperService(delay)
        }
    }

    fun scheduleChangeWallpaperService(delay: Long) {
        shutdown()
        schedule(mChangeWallpaperRunnable, delay, TimeUnit.MILLISECONDS)
    }

    fun schedule(runnable: Runnable, delay: Long, timeUnit: TimeUnit) {
        shutdown()
        mScheduledExecutorService
                .scheduleAtFixedRate(runnable, delay, delay, timeUnit)
    }
}
